compiler/src/fuzzer/input_pool.rs

```diff
@@ -1,44 +1,62 @@
+use std::{cell::RefCell, rc::Rc};
+
+use itertools::Itertools;
 use rand::{rngs::ThreadRng, seq::SliceRandom, Rng};
+use rustc_hash::FxHashMap;
+
+use crate::vm::{Heap, Pointer};
 
 use super::{
-    utils::Input,
+    input::Input,
     values::{generate_input, mutate_input},
 };
 
 pub type Score = f64;
 
 pub struct InputPool {
+    heap: Rc<RefCell<Heap>>,
     num_args: usize,
-    symbols: Vec<String>,
-    inputs_and_scores: Vec<(Input, Score)>,
+    symbols: Vec<Pointer>,
+    input_scores: FxHashMap<Input, Score>,
 }
 
 impl InputPool {
     pub fn new(num_args: usize, symbols_in_heap: Vec<String>) -> Self {
+        let mut heap = Heap::default();
+
+        let mut symbols = symbols_in_heap
+            .into_iter()
+            .map(|symbol| heap.create_symbol(symbol))
+            .collect_vec();
+        if symbols.is_empty() {
+            symbols.push(heap.create_symbol("Nothing".to_string()));
+        }
+
         Self {
+            heap: Rc::new(RefCell::new(heap)),
             num_args,
-            symbols: symbols_in_heap,
-            inputs_and_scores: vec![],
+            symbols,
+            input_scores: FxHashMap::default(),
         }
     }
 
     pub fn generate_new_input(&self) -> Input {
         let mut rng = ThreadRng::default();
```

Comment:
AFAICT, this still doesn't check whether the new input already exists in `self.input_scores`

Reply:
Oh, right 😅